src/api/app/models/workflow/step/branch_package_step.rb

```diff
@@ -0,0 +1,138 @@
+class Workflow
+  module Step
+    class BranchPackageStep
+      include ActiveModel::Model
+
+      attr_reader :errors
+
+      validates :source_project, :source_package, presence: true
+
+      def initialize(step_instructions:, scm_extractor_payload:)
+        @step_instructions = step_instructions
+        @scm_extractor_payload = scm_extractor_payload
+        @errors = []
+      end
+
+      def allowed_event_and_action?
+        new_pull_request?
+        # TODO: new_pull_request? || updated_pull_request? || closed_pull_request?
+      end
+
+      def call
+        # Decide what to do by action:
+        # - opened -> create a branch package on OBS using the configuration file's details and the PR number.
+        return unless new_pull_request?
+
+        branched_package = branch
+        add_branch_request_file(package: branched_package)
+
+        branched_package
+        # NICE TO HAVE:
+        # - synchronize -> get the existent branched package on OBS and ensure it updates the source
+        #                  code and it rebuilds (trigger service). We should have a previous branch with the
+        #                  contents of the initial pull_request or the previous synchronization.
+        # - closed -> remove the existent branched package
+      end
+
+      private
+
+      def source_project
+        @step_instructions['source_project']
+      end
+
+      def source_package
+        @step_instructions['source_package']
+      end
+
+      def destination_package
+        return @step_instructions['destination_package'] if @step_instructions['destination_package'].present?
+
+        source_package
+      end
+
+      def destination_project
+        source_project + ":PR-#{@scm_extractor_payload[:pr_number]}"
+      end
+
+      def remote_source?
+        return true if Project.find_remote_project(source_project)
+
+        false
+      end
+
+      def check_source_access
+        return if remote_source?
+
+        options = { use_source: false, follow_project_links: true, follow_multibuild: true }
+        src_package = Package.get_by_project_and_name(source_project, source_package, options)
+
+        raise Pundit::NotAuthorizedError unless PackagePolicy.new(User.session, src_package).create_branch?
+      end
+
+      def branch
+        check_source_access
```

Comment:
Why only source? What about the destination? Or do you want to rely on the checks inside `BranchPackage`?

Reply:
Yes the `BranchPackage` model checks for the destination, so the idea was to rely on them, since that is the same way its done in the controller endpoints for package branching.